lookout/core/lib.py

```diff
@@ -87,38 +91,83 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
     return result
 
 
-def filter_files(files: Dict[str, File], line_length_limit: int, overall_size_limit: int,
-                 random_state: int = 7, log: Optional[logging.Logger] = None) -> List[File]:
+def filter_filepaths(filepaths: Iterable[str], exclude_pattern: Optional[str] = None,
+                     ) -> Iterable[str]:
     """
-    Filter files based on their maximum line length.
+    Mirror of the file filtering used in the format analyzer for use by debugging tools.
 
-    :param files: files to filter.
+    :param filepaths: Iterable of filepaths to filter.
+    :param exclude_pattern: Pattern to reject files based on their path. If None, uses the pattern
+                            currently in use in lookout.core. Use "" to not filter anything.
+    :return Iterable of paths, filtered.
+    """
+    if exclude_pattern is None:
+        exclude_pattern = GARBAGE_PATTERN
+    exclude_compiled_pattern = re.compile(exclude_pattern) if exclude_pattern else None
+    for filepath in filepaths:
+        if not isfile(filepath):
+            continue
+        if exclude_compiled_pattern and exclude_compiled_pattern.search(filepath):
+            continue
+        yield filepath
+
+
+def parse_files(filenames: Iterable[str], line_length_limit: int, overall_size_limit: int,
+                client: BblfshClient, language: str, random_state: int = 7,
+                progress_tracker: Callable = lambda x: x, log: Optional[logging.Logger] = None
+                ) -> Iterator[File]:
+    """
+    Filter files based on `language` and their maximum line length.
+
+    :param filenames: paths to the files to filter.
     :param line_length_limit: maximum line length to accept a file.
+    :param client: Babelfish client. Babelfish server should be started accordingly.
+    :param language: Language to consider. Will discard the other languages.
     :param overall_size_limit: maximum cumulative files size in bytes. \
                                The files are discarded after reaching this limit.
     :param random_state: random generator state for shuffling the files.
+    :param progress_tracker: Optional progress metric whenn iterating over the input files.
     :param log: logger to use to report the number of excluded files.
     :return: files passed through the filter and the number of files which were excluded.
     """
+    n_parsed = 0
     line_passed = []
-    for key, file in files.items():
-        if len(max(file.content.splitlines(), key=len, default=b"")) <= line_length_limit:
-            line_passed.append(key)
+    n_filenames_filtered = len(list(filter_filepaths(filenames)))
+    for filename in progress_tracker(filter_filepaths(filenames)):
```

Comment:
We don't want to parse everything just to realize that we drop some fraction due to `overall_size_limit`